### PR TITLE
taint analysis: Improve handling of dataflow for tainted value propagation in class field definitions

### LINKED ISSUE #7742: Kotlin does not detect flow of string inside Objects
**Describe the bug**
Semgrep appears to ignore dataflow inside objects https://semgrep.dev/s/JKYj 

```
object Hi {
    private const val MYSTRING = "123"

    // should match
    private val magicString = Base64.decode(MYSTRING, Base64.DEFAULT).toString(Charsets.UTF_8)

}
```

Is ignored from the pattern in the rule above

**To Reproduce**
https://semgrep.dev/s/JKYj

**Expected behavior**
The rule should match inside objects too

**What is the priority of the bug to you?**

- [x] P2: regular bug that should get fixed

**Environment**
If not using semgrep.dev: are you running off docker, an official binary, a local build?

**Use case**
Match more findings in Kotlin

### PR DESCRIPTION
**Pull request by ihji**
This commit resolves an issue where dataflow was not correctly accounted for when tainted values flowed through field definitions in class definitions. In JVM-based programming languages such as Java, Kotlin, and Scala, the order of field definitions is significant, and they function like statements during object initialization. To address this, during the transformation from AST to IL, we now treat a list of field definitions as a sequence of statements, allowing for proper tracking of tainted values during object initialization.

Closes #7742

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by IagoAbal** [comment]
Is there a ticket associated with this? Shouldn't we add a changelog entry too?

File: tests/tainting_rules/kotlin/tainted_args.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+rules:
```
Comment (IagoAbal) [inline tests/tainting_rules/kotlin/tainted_args.yaml]: Sorry for the confusion, `tests/tainting_rules` should not be used anymore... It exists for historical reasons. Need to clean that up. Taint tests now go under `tests/rules/taint_*`.

File: src/analyzing/AST_to_IL.ml [hunk 2]
```diff
@@ -1241,6 +1241,9 @@ and stmt_aux env st =
       let ss, e' = expr_with_pre_stmts env e in
       let lv = lval_of_ent env ent in
       ss @ [ mk_s (Instr (mk_i (Assign (lv, e')) (Related (G.S st)))) ]
+  | G.DefStmt (_, G.ClassDef { cbody = fields; _ }) ->
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: In general I don't think this is correct, the fields could be modified by calls to methods no? It's correct if the only thing you're analyzing are the fields themselves. It would be useful to get more context here.

Comment (ihji) [reply src/analyzing/AST_to_IL.ml]: Certainly, it is possible to modify the fields after initialization. However, the primary objective of this PR is to perform data flow analysis specifically during object initialization. The purpose is to address issue #7742 by resolving a limitation that cannot be resolved using AST alone. In the AST, each field in the object definition is treated independently, but in taint mode, there is a partial feasibility in handling this scenario.

However, since I lack sufficient context on the design decisions made during IL transformation, I am also open to discarding these changes if the negative side effects outweigh the positives.

Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml]: Then IMO we should have a special visiting step that just analyzes fields in `Match_tainting_mode.ml`. ~Although I think this could be solved already with sym-prop. The #7742's case should be even simpler because it only needs const-prop.~ However when I open #7742 and I follow the Playground link, what I see is a search-mode rule.

Comment (ihji) [reply src/analyzing/AST_to_IL.ml]: The author of #7742 attempted to perform semi-taint analysis in search mode. However, this approach would not work with class fields without significant modifications to the class_definition AST. The PR enables the same analysis in taint mode, which appears to be more suitable for this particular case.

File: tests/rules/tainted_anon_type.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+rules:
```
Comment (IagoAbal) [inline tests/rules/tainted_anon_type.yaml]: `tests/rules/taint_...` instead of `tainted_....`, there are a few tests that break the convention already, no big deal but FYI.

File: changelog.d/gh-7742.fixed [hunk 4]
```diff
@@ -0,0 +1 @@
+taint analysis: Improve handling of dataflow for tainted value propagation in class field definitions
```
Comment (IagoAbal) [inline changelog.d/gh-7742.fixed:1]: We should do our best so that changelog entries are easy to understand by e.g. technical writers. Not always easy, but to keep in mind. If I didn't know what is this about, and I just read that changelog entry, I wouldn't really know what had changed. From the user perspective, what works now that didn't work before?

Comment (IagoAbal) [reply changelog.d/gh-7742.fixed:1]: cc @adamkvitek

File: tests/rules/tainted_anon_type.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class Asdf {
+    class ReturnObject1 {
+        public string Secret;
+    }
+
+    public void Test1() {
+        // ruleid: tainted-anon-type
+        Log(new ReturnObject1() { Secret = GetSecret() });
```
Comment (IagoAbal) [inline tests/rules/tainted_anon_type.cs]: No problem to have it here, but this test seems unrelated to the change, am I correct?

**Review by IagoAbal — APPROVED** [review]

File: src/analyzing/AST_to_IL.ml (hunk 2 above)
Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml]: I'm not so sure about that, have you actually talked to Lewis about it?

Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml]: I am happy to merge this, I am not so sure you are actually fixing #7742.

File: tests/rules/tainted_anon_type.yaml (hunk 3 above)
Comment (ihji) [reply tests/rules/tainted_anon_type.yaml]: renamed.

File: changelog.d/gh-7742.fixed (hunk 4 above)
Comment (ihji) [reply changelog.d/gh-7742.fixed:1]: amended the change description.

File: tests/rules/tainted_anon_type.cs (hunk 5 above)
Comment (ihji) [reply tests/rules/tainted_anon_type.cs]: I included the test case because it appeared that this PR might also address #7995, but it turned out that it only works for scenarios where the source and target are in the same expression.

Removed the anonymous type test case.

**Comment by IagoAbal** [comment]
If you add
```
Closes #7742
```
to your PR comment, we keep traceability and GitHub automatically closes the ticket when this is merged.

File: changelog.d/gh-7742.fixed [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+This change resolves an issue where dataflow was not correctly accounted for
```
Comment (IagoAbal) [inline changelog.d/gh-7742.fixed:3]: This is quite similar to the PR comment which is oriented towards other PA members. This should be directed to non-PA people, so IMO e.g. implementation details aren't needed here. The last sentence is what I would personally put first, "Track tainted values during object initialization" or similar, followed by a description of what stuff now works "Semgrep will now track the flow of taint sources into sinks during object initialization" or similar. It doesn't need to be very long, in fact I would try to keep it short.  

Sorry for nit picking, I am not trying to be pedantic. Of course, I am not necessarily right, but we could check with Bruno and Adam for a second opinion. Both of them have complained in the past about the quality of the changelog entries. I'm just trying to push a bit so that over time we write better changelogs!
